File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxProduct = function(head) {
    const totalSum = (function dfsSum(node) {
        if (!node) return 0;
        return dfsSum(node.left) + dfsSum(node.right) + node.val;
    })(head);
    
    let result = 0;
    const modulo = 1000000007;
    
    const dfs = (node) => {
        if (!node) return 0;
        const left = dfs(node.left);
        const right = dfs(node.right);
        result = Math.max(result, (totalSum - left) * left, (totalSum - right) * right);
        return left + right + node.val;
    }
    dfs(head);
    return result % modulo;
};

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        totalSum = 0
        result = 0
        modulo = (10 ** 9) + 7

        def _dfs(node, caculate=False) -> int:
            nonlocal result
            l = 0
            r = 0

            if node.left:
                left = _dfs(node.left, caculate) 
                l = left if left is not None else 0
            if node.right:
                right = _dfs(node.right, caculate)
                r = right if right is not None else 0

            curSum = (l + r + node.val)
            if caculate:
                multiply = ((totalSum - curSum) * curSum) 
                result = max(multiply, result)

            return curSum 

        totalSum = _dfs(root)
        _dfs(root, True)
        return result % modulo